psd-web/app/controllers/concerns/secondary_authentication_concern.rb

```diff
@@ -0,0 +1,62 @@
+# By default is using 'secondary_authentication' operation.
+# To protect some actions with custom `secondary_authentication`,
+# please override `user_id_for_secondary_authentication` and `current_operation` methods
+# in such controller
+#
+# Only one action in controller can be protected by secondary authentication.
+module SecondaryAuthenticationConcern
+  extend ActiveSupport::Concern
+
+  def require_secondary_authentication
+    return unless Rails.configuration.secondary_authentication_enabled
+
+    perform_secondary_authentication
+  end
+
+  def perform_secondary_authentication
+    if user_id_for_secondary_authentication && !secondary_authentication_present_for_operation_and_user
+      session[:secondary_authentication_redirect_to] = request.fullpath
+      auth = SecondaryAuthentication.create(user_id: user_id_for_secondary_authentication, operation: current_operation)
+      auth.generate_and_send_code
+      redirect_to new_secondary_authentications_path(secondary_authentication_id: auth.id)
+    end
+  end
+
+  # Use as `before_filter` in application_controller controller
+  def ensure_secondary_authentication
+    session[:secondary_authentication] ||= []
+  end
+
+  # used in application controller to cleanup old auths
+  def cleanup_secondary_authentication
+    session[:secondary_authentication].reject! do |auth_id|
+      unless SecondaryAuthentication.find_by(id: auth_id)
+        return true
+      end
+
+      if SecondaryAuthentication.find(auth_id).expired?
+        SecondaryAuthentication.find(auth_id).delete
+        true
+      else
+        false
+      end
```

Comment:
Suggestion: Seems we're hitting the DB 3 times here for retrieving the same information. 
Can't we just get the `SecondaryAuthentication` instance once and operate over it? Or there was a explicit intent behind it (that I'm obviously missing)

```suggestion
      secondary_authentication = SecondaryAuthentication.find_by(id: auth_id)
      return true unless secondary_authentication

      if secondary_authentication.expired?
        secondary_authentication.delete
        true
      else
        false
      end
```